File: go/src/907.子数组的最小值之和.go
/*
 * @lc app=leetcode.cn id=907 lang=golang
 *
 * [907] 子数组的最小值之和
 */

// @lc code=start
func sumSubarrayMins(A []int) int {
	// 利用单调栈
	mod := 1e9 + 7
	size, res := len(A), 0
	left, right := make([]int, size), make([]int, size)

	// 找到该元素左右两边 不比 该元素大的元素的下标
	stack := []int{}
	for i := 0; i < size; i++ {
		for len(stack) > 0 && A[stack[len(stack) - 1]] >= A[i] {
			stack = stack[:len(stack) - 1]
		}

		if len(stack) == 0 {
			left[i] = -1
		} else {
			left[i] = stack[len(stack) - 1]
		}

		stack = append(stack, i)
	}

	stack = []int{}
	for i := size - 1; i >= 0; i-- {
		for len(stack) > 0 && A[stack[len(stack) - 1]] > A[i] {
			stack = stack[:len(stack) - 1]
		}

		if len(stack) == 0 {
			right[i] = size
		} else {
			right[i] = stack[len(stack) - 1]
		}

		stack = append(stack, i)
	}

	for i := 0; i < size; i++ {
		res += A[i] * (right[i] - i) * (i - left[i])
	}
	return res % int(mod)
}


func sumSubarrayMins2(A []int) int {
	mod := 1e9 + 7
	size, res := len(A), 0
	left, right := make([]int, size), make([]int, size)

	// 找到该元素左右两边 不比 该元素大的元素的下标
	for i := 0; i < size; i++ {
		j := i - 1
		for j >= 0 && A[j] >= A[i] {
			j--
		} 
		left[i] = j

		j = i + 1
		for j < size && A[j] > A[i] {
			j++
		} 
		right[i] = j
	}	

	for i := 0; i < size; i++ {
		res += A[i] * (right[i] - i) *(i - left[i])
	}
	return res % int(mod)
}
// @lc code=end

File: rust/src/907.子数组的最小值之和.rs
/*
 * @lc app=leetcode.cn id=907 lang=rust
 *
 * [907] 子数组的最小值之和
 */

// @lc code=start
impl Solution {
    pub fn sum_subarray_mins(arr: Vec<i32>) -> i32 {
        const MOD: usize = 1e9 as usize + 7;
        let n = arr.len();
        let (mut left, mut right) = (vec![0i32; n], vec![0i32; n]);
        // 找到该元素左右两边 不比 该元素大的元素的下标
        let mut stack = vec![];
        for i in 0..n {
            while stack.len() > 0 && arr[*stack.last().unwrap() as usize] >= arr[i] {
                stack.pop();
            }

            left[i] = if stack.len() == 0 {
                -1
            } else {
                *stack.last().unwrap()
            };

            stack.push(i as i32);
        }

        let mut stack = vec![];
        for i in (0..n).rev() {
            while stack.len() > 0 && arr[*stack.last().unwrap() as usize] > arr[i] {
                stack.pop();
            }

            right[i] = if stack.len() == 0 {
                n as i32
            } else {
                *stack.last().unwrap()
            };

            stack.push(i as i32);
        }

        let mut res = 0;
        for i in 0..n {
            res = (res
                + arr[i] as usize * (i as i32 - left[i]) as usize * (right[i] - i as i32) as usize)
                % MOD;
        }

        res as i32
    }
}
// @lc code=end
